File: src/v2/domain/ports/pending_state_port.py
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass, field

from src.v2.domain.entities.expense import ExtractedExpense

_TTL_SECONDS = 600  # 10 minutes


@dataclass
class PendingExpense:
    extracted: ExtractedExpense
    category: str          # display name shown in confirmation message
    category_id: int       # used by ConfirmExpense to persist
    chat_id: int
    message_id: int | None = None
    expires_at: float = field(
        default_factory=lambda: time.monotonic() + _TTL_SECONDS
    )

    def is_expired(self) -> bool:
        return time.monotonic() > self.expires_at


class PendingStatePort(ABC):
    @abstractmethod
    def set(self, chat_id: int, state: PendingExpense) -> None: ...

    @abstractmethod
    def get(self, chat_id: int) -> PendingExpense | None: ...

    @abstractmethod
    def update_category(
        self, chat_id: int, category: str, category_id: int
    ) -> bool:
        """Update category on an existing pending state. Returns False if expired/missing."""
        ...

    @abstractmethod
    def clear(self, chat_id: int) -> None: ...

File: app/src/main/java/br/com/nathanfiorito/finances/domain/telegram/ports/PendingStatePort.java
package br.com.nathanfiorito.finances.domain.telegram.ports;

import br.com.nathanfiorito.finances.domain.telegram.records.PendingTransaction;

import java.util.Optional;

public interface PendingStatePort {

    void set(long chatId, PendingTransaction state);

    Optional<PendingTransaction> get(long chatId);

    /**
     * Updates the category on an existing pending state.
     *
     * @return false if the state is missing or expired
     */
    boolean updateCategory(long chatId, String category, int categoryId);

    void clear(long chatId);
}
